src/openloco/windows/KeyboardShortcuts.cpp

```diff
@@ -0,0 +1,200 @@
+        _events.get_scroll_size = get_scroll_size;
+        _events.scroll_mouse_down = on_scroll_mouse_down;
+        _events.scroll_mouse_over = on_scroll_mouse_over;
+        _events.tooltip = tooltip;
+        _events.draw = draw;
+        _events.draw_scroll = draw_scroll;
+    }
+
+    // 0x004BE6C7
+    window* open()
+    {
+        window* window;
+
+        window = WindowManager::bringToFront(WindowType::keyboardShortcuts, 0);
+        if (window != nullptr)
+            return window;
+
+        init_events();
+
+        // 0x004BF833 (create_options_window)
+        window = WindowManager::createWindowCentred(WindowType::keyboardShortcuts, { 360, 238 }, 0, &_events);
+
+        window->widgets = _widgets;
+        window->enabled_widgets = (1 << widx::close_button) | (1 << widx::customize_keys_btn);
+        window->init_scroll_widgets();
+
+        auto skin = objectmgr::get<interface_skin_object>();
+        window->colours[0] = skin->colour_0B;
+        window->colours[1] = skin->colour_10;
+
+        window->row_count = 35;
+        window->row_hover = -1;
+
+        return window;
+    }
+
+    // 0x004BE726
+    static void draw(ui::window* window, gfx::drawpixelinfo_t* dpi)
+    {
+        // Draw widgets.
+        window->draw(dpi);
+    }
+    // 0x004BE72C
+    static void draw_scroll(ui::window* window, gfx::drawpixelinfo_t* dpi, uint32_t scrollIndex)
+    {
+        auto colour = window->colours[1];
+        auto shade = colour::get_shade(colour, 4);
+        gfx::clear_single(*dpi, shade);
+
+        auto dx = 0;
+        for (auto i = 0; i < window->row_count; i++)
+        {
+            string_id ax = string_ids::white_stringid2;
+            if (i == window->row_hover)
+            {
+                gfx::draw_rect(dpi, 0, dx, 800, rowHeight, 0x2000030);
+                ax = string_ids::wcolour2_stringid2;
+            }
+
+            addr<0x112C826 + 2, string_id>() = string_ids::shortcut_close_topmost_window + i; //string id
+            addr<0x112C826 + 4, string_id>() = string_ids::empty;
+            addr<0x112C826 + 6, string_id>() = string_ids::empty;
+
+            if (config::get().keyboard_shortcuts[i].var_0 != 0xFF)
+            {
+                addr<0x112C826 + 6, string_id>() = string_ids::str_740 + config::get().keyboard_shortcuts[i].var_0;
+
+                if (config::get().keyboard_shortcuts[i].var_1 != 0)
+                {
+                    addr<0x112C826 + 4, string_id>() = string_ids::keyboard_shortcut_modifier_shift;
+                    if (config::get().keyboard_shortcuts[i].var_1 != 1)
+                    {
+                        addr<0x112C826 + 4, string_id>() = string_ids::keyboard_shortcut_modifier_ctrl;
+                    }
+                }
+            }
+
+            addr<0x112C826 + 0, string_id>() = string_ids::keyboard_shortcut_list_format;
+
+            gfx::draw_string_494B3F(*dpi, 0, dx - 1, colour::black, ax, _commonFormatArgs);
+            dx += rowHeight;
```

Comment:
To do: continue/break out early when drawing outside the visible scrollview area.

Reply:
The original does not seem to be doing this either here. Perhaps something we introduced in OpenRCT2 as an optimisation?

Reply:
Likely. Though I don't see the need to optimise such a not-frequently used window. I think it'd make more sense to optimise once we revisit the widget system, and call a render function per row. That'd optimise every scrollview.